File: src/main/java/org/capturecoop/cccolorutils/chooser/CCColorChooser.java
package org.capturecoop.cccolorutils.chooser;

import org.capturecoop.cccolorutils.CCColor;
import org.capturecoop.cccolorutils.chooser.gui.CCColorChooserPreviewPanel;
import org.capturecoop.cccolorutils.chooser.gui.CCColorChooserSetterPanel;
import org.capturecoop.ccutils.utils.ICCClosable;

import javax.swing.*;
import javax.swing.event.ChangeEvent;
import javax.swing.event.ChangeListener;
import java.awt.*;
import java.awt.image.BufferedImage;
import java.util.ArrayList;

public class CCColorChooser extends JFrame implements ICCClosable {
    private final CCColorChooser instance;
    private CCColor color;
    private final CCColorChooserPreviewPanel previewPanel;
    private final CCColorChooserSetterPanel setterPanel;

    private final ArrayList<ChangeListener> changeListeners = new ArrayList<>();
    private final ChangeListener colorChangeListener;

    public CCColorChooser(CCColor color, String title, int x, int y, boolean useGradient, BufferedImage backgroundImage, BufferedImage icon) {
        instance = this;
        this.color = color;
        setterPanel = new CCColorChooserSetterPanel(color.getPrimaryColor(), this);
        previewPanel = new CCColorChooserPreviewPanel(this, useGradient, backgroundImage);

        colorChangeListener = e -> alertChangeListeners();
        color.addChangeListener(colorChangeListener);

        setTitle(title);
        if(icon != null) setIconImage(icon);
        init(x, y);
    }

    public void init(int x, int y) {
        JPanel mainPanel = new JPanel();
        JPanel submitButtonPanel = new JPanel();
        JButton submit = new JButton("Okay");
        submit.addActionListener(e -> instance.close());

        mainPanel.setLayout(new BoxLayout(mainPanel, BoxLayout.Y_AXIS));

        mainPanel.add(previewPanel);
        mainPanel.add(setterPanel);
        mainPanel.add(submitButtonPanel);

        add(mainPanel);
        setFocusable(true);

        setAlwaysOnTop(true);

        pack();
        submit.setPreferredSize(new Dimension(this.getWidth()/2, 50));
        submitButtonPanel.add(submit);

        pack();
        setLocation(x - getWidth()/2, y - getHeight()/2);
        setVisible(true);
    }

    public CCColorChooserSetterPanel getSetterPanel() {
        return setterPanel;
    }

    public CCColor getColor() {
        return color;
    }

    public void addChangeListener(ChangeListener listener) {
        changeListeners.add(listener);
    }

    public void removeChangeListener(ChangeListener listener) {
        changeListeners.remove(listener);
    }

    public void alertChangeListeners() {
        for(ChangeListener listener : changeListeners) {
            listener.stateChanged(new ChangeEvent(this));
        }
    }

    @Override
    public void close() {
        color.removeChangeListener(colorChangeListener);
        dispose();
    }
}

File: src/main/kotlin/org/capturecoop/cccolorutils/chooser/CCColorChooser.kt
package org.capturecoop.cccolorutils.chooser

import org.capturecoop.cccolorutils.CCColor
import org.capturecoop.cccolorutils.chooser.gui.CCColorChooserPreviewPanel
import org.capturecoop.cccolorutils.chooser.gui.CCColorChooserSetterPanel
import org.capturecoop.ccutils.utils.CCIClosable
import java.awt.Color
import java.awt.Dimension
import java.awt.Point
import java.awt.Toolkit
import java.awt.event.WindowAdapter
import java.awt.event.WindowEvent
import java.awt.image.BufferedImage
import javax.swing.BoxLayout
import javax.swing.JButton
import javax.swing.JFrame
import javax.swing.JPanel
import javax.swing.event.ChangeEvent
import javax.swing.event.ChangeListener

class CCColorChooser(color: CCColor = CCColor(Color.WHITE), title: String = "Color Chooser", parent: JFrame? = null, x: Int = -Integer.MAX_VALUE, y: Int = -Integer.MAX_VALUE, useGradient: Boolean = color.isGradient, backgroundImage: BufferedImage? = null, icon: BufferedImage? = null): JFrame(), CCIClosable {
    var color: CCColor = color
        set(value) {
            color.loadFromCCColor(value)
            //Update setter, preview panel updates automatically via a listener in cccolor
            setterPanel.setColor(color.primaryColor, alertListeners = true, updateComponents = true)
        }
    private val previewPanel: CCColorChooserPreviewPanel
    val setterPanel: CCColorChooserSetterPanel
    private val changeListeners = ArrayList<ChangeListener>()
    private val colorChangeListener: ChangeListener
    private var onClose: ((CCColorChooser) -> (Unit))? = null

    init {
        setterPanel = CCColorChooserSetterPanel(color.primaryColor, this)
        previewPanel = CCColorChooserPreviewPanel(this, useGradient, backgroundImage)
        colorChangeListener = ChangeListener { alertChangeListeners() }
        color.addChangeListener(colorChangeListener)
        setTitle(title)
        if (icon != null) iconImage = icon
        init(x, y, parent)
    }

    private fun init(x: Int, y: Int, parent: JFrame?) {
        val mainPanel = JPanel()
        val submitButtonPanel = JPanel()
        val submit = JButton("Okay")
        submit.addActionListener { close() }
        mainPanel.layout = BoxLayout(mainPanel, BoxLayout.Y_AXIS)
        mainPanel.add(previewPanel)
        mainPanel.add(setterPanel)
        mainPanel.add(submitButtonPanel)
        add(mainPanel)
        isFocusable = true
        isAlwaysOnTop = true
        pack()
        submit.preferredSize = Dimension(width / 2, 50)
        submitButtonPanel.add(submit)
        pack()
        if(parent == null) {
            Toolkit.getDefaultToolkit().screenSize.also {
                val posX = if(x != -Integer.MAX_VALUE) x else it.width / 2 - width / 2
                val posY = if(y != -Integer.MAX_VALUE) y else it.height / 2 - height / 2
                location = Point(posX, posY)
            }
        } else {
            location = Point(parent.location.x + parent.width / 2 - width / 2, parent.location.y + parent.height  / 2 - height / 2)
        }
        isVisible = true
        addWindowListener(object: WindowAdapter() {
            override fun windowClosing(e: WindowEvent?) {
                close()
            }
        })
    }

    private fun alertChangeListeners() {
        changeListeners.forEach { it.stateChanged(ChangeEvent(this)) }
    }

    fun setOnClose(action: (CCColorChooser) -> (Unit)) {
        onClose = action
    }

    override fun close() {
        onClose?.invoke(this)
        color.removeChangeListener(colorChangeListener)
        dispose()
    }
}